crates/bevy_ecs/src/storage/thin_array_ptr.rs

```diff
@@ -0,0 +1,250 @@
+                self.get_unchecked_raw(index),
+                self.get_unchecked_raw(last_element_index),
+                1,
+            );
+        }
+        self.get_unchecked_raw(last_element_index)
+    }
+
+    // TODO: Docs
+    /// # Safety
+    /// The caller must:
+    /// - ensure that `index < len`
+    /// - ensure that `last_element_index` = `len - 1`
+    /// - update their saved length value to reflect that the last element has been removed (decrement it)
+    pub unsafe fn swap_remove_and_drop_unchecked(
+        &mut self,
+        index: usize,
+        last_element_index: usize,
+    ) {
+        std::ptr::drop_in_place(self.swap_remove_and_forget_unchecked(index, last_element_index))
+    }
+
+    /// Pop the last element of the array.
+    ///
+    /// # Safety
+    /// - ensure that `current_len` is indeed the len of the array
+    /// - update their saved `len` (decrement it)
+    pub unsafe fn pop(&mut self, current_len: usize) -> Option<T> {
+        if current_len == 0 {
+            None
+        } else {
+            Some(std::ptr::read(self.data.as_ptr().add(current_len - 1)))
+        }
+    }
+
+    /// Clears the array, removing (and dropping) Note that this method has no effect on the allocated capacity of the vector.
+    ///
+    /// # Safety
+    /// The caller must:
+    /// - ensure that `current_len` is indeed the length of the array
+    /// - update their saved length value
+    pub unsafe fn clear_elements(&mut self, mut current_len: usize) {
+        while self.pop(current_len).is_some() {
+            current_len -= 1;
+        }
+    }
+
+    /// Drop the entire array and all its elements.
+    ///
+    /// # Safety
+    /// The caller must:
+    /// - ensure that `current_len` is indeed the length of the array
+    /// - ensure that `current_capacity` is indeed the capacity of the array
+    pub unsafe fn drop(&mut self, current_capacity: usize, current_len: usize) {
+        if current_capacity != 0 {
+            self.clear_elements(current_len);
+            let layout = Layout::array::<T>(current_capacity).expect("layout should be valid");
+            std::alloc::dealloc(self.data.as_ptr().cast(), layout);
+        }
+    }
+
+    /// Get the [`ThinArrayPtr`] as a slice with a given length.
+    ///
+    /// # Safety
+    /// - `slice_len` must match the actual length of the array
+    /// but if `slice_len` will be smaller, the slice will just be smaller than need be - no UB
+    pub unsafe fn as_slice<'a>(&'a self, slice_len: usize) -> &'a [T] {
+        // SAFETY:
+        // - the data is valid - allocated with the same allocater
+        // - non-null and well-aligned
+        // - we have a shared refernce to self - the data will not be mutated during 'a
+        unsafe { std::slice::from_raw_parts(self.data.as_ptr(), slice_len) }
+    }
+}
+
+impl<T> From<Box<[T]>> for ThinArrayPtr<T> {
+    fn from(value: Box<[T]>) -> Self {
+        if Layout::new::<T>().size() == 0 {
```

Comment:
If you wrap this in a const (i.e. `const _: () = { ... }`, we could make this a const panic, and force this to not compile.